File: exercise/28/kmp.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        if (needle.empty()) {
            return 0;
        } else if (haystack.empty()) {
            return -1;
        }
        int m = haystack.length(), n = needle.length();
        int i = 0, j = -1;
        vector<int> next(n + 1);
        next[0] = -1;
        while (i < n) {
            if (j == -1 || needle[i] == needle[j]) {
                i++, j++;
                next[i] = j;
            } else {
                j = next[j];
            }
        }
        i = j = 0;
        while (i < m && j < n) {
            if (j == -1 || haystack[i] == needle[j]) {
                i++, j++;
            } else {
                j = next[j];
            }
        }
        if (j == n) {
            return i - j;
        } else {
            return -1;
        }
    }
};

File: exercise/28/kmp.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        m = len(haystack)
        n = len(needle)
        if n == 0:
            return 0
        if n > m:
            return -1
        move = [0 for _ in range(n + 1)]
        move[0] = -1
        i, j = 0, -1
        while i <= n and j <= n:
            if j == -1 or needle[i] == needle[j]:
                i += 1
                j += 1
                move[i] = j
            else:
                j = move[j]
        i, j = 0, 0
        while i < m and j < n:
            if j == -1 or haystack[i] == needle[j]:
                i += 1
                j += 1
            else:
                j = move[j]
        if j == n:
            return i - j
        else:
            return -1
